### PR TITLE
NXP-30864: Implement CI build pipeline for nuxeo-compound-documents

### PR DESCRIPTION
**Pull request by tfowley**

### CODE REVIEW

File: ci/helm/values/values-elasticsearch.yaml.gotmpl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+persistence:
+  enabled: false
+tolerations:
+  - key: team
+    operator: Equal
+    value: platform
+    effect: NoSchedule
+nodeSelector:
+  team: platform
+
```
Comment (troger) [inline ci/helm/values/values-elasticsearch.yaml.gotmpl]: Empty line.

**Review by troger — COMMENTED** [review]
👍

File: ci/Jenkinsfiles/build.groovy [hunk 2]
```diff
@@ -0,0 +1,445 @@
+/*
+* (C) Copyright 2021 Nuxeo (http://nuxeo.com/) and others.
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: 2022, detail though

File: ci/Jenkinsfiles/build.groovy [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+  return {
+    stage("Run ${env} unit tests") {
+      container("${containerName}") {
+        // TODO NXP-29512: on a PR, make the build continue even if there is a test error
+        // on other environments than the dev one
+        // to remove when all test environments will be mandatory
+        catchError(buildResult: isPullRequest() && "dev" != env ? 'SUCCESS' : 'FAILURE', stageResult: 'FAILURE', catchInterruptions: false) {
+          script {
+            setGitHubBuildStatus("utests/${env}", "Unit tests - ${env} environment", 'PENDING')
+            sh "kubectl create namespace ${testNamespace}"
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: surround with `if (!isDev)`, we don't need it in dev

File: ci/Jenkinsfiles/build.groovy [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+            try {
+              echo """
+              ----------------------------------------
+              Run ${env} unit tests
+              ----------------------------------------"""
+
+              echo "${env} unit tests: run Maven"
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: this should in fact be displayed just before running the Maven command, above `sh "${mvnCommand}"`

File: ci/Jenkinsfiles/build.groovy [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+  }
+}
+
+void archiveKafkaLogs(namespace, logFile) {
+  // don't fail if pod doesn't exist
+  sh "kubectl logs ${TEST_KAFKA_POD_NAME} --namespace=${namespace} > ${logFile} || true"
+  archiveArtifacts allowEmptyArchive: true, artifacts: "${logFile}"
+}
+
+void gitCheckout(String reference, String option = '') {
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: unused

File: ci/Jenkinsfiles/build.groovy [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+  sh """
+    git checkout ${option} ${reference}
+  """
+}
+
+void gitPush(String reference) {
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: unused

File: ci/Jenkinsfiles/build.groovy [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+  sh """
+    git push origin ${reference}
+  """
+}
+
+void gitTag(String tagname, String message) {
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: unused

File: ci/Jenkinsfiles/build.groovy [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+    git tag -a ${tagname} -m "${message}"
+  """
+}
+
+pipeline {
+  agent {
+    label 'jenkins-nuxeo-platform-lts-2021'
+  }
+  options {
+    disableConcurrentBuilds()
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: We sould follow what was done for nxueo-lts with https://jira.nuxeo.com/browse/NXP-29543: https://github.com/nuxeo/nuxeo-lts/commit/e8cbf7e0b155f7b797cab33853d7eb9e7429a1aa
So, remove this call to `disableConcurrentBuilds` and add the `abortRunningBuilds` block at the very beigging of the pipeline.

File: ci/Jenkinsfiles/build.groovy [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+    buildDiscarder(logRotator(daysToKeepStr: '15', numToKeepStr: '10', artifactNumToKeepStr: '5'))
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: As in https://github.com/nuxeo/nuxeo-lts/blob/2021/ci/Jenkinsfiles/build.groovy#L41., we usually use:
- daysToKeepStr: 60 # days to keep builds
- numToKeepStr: 60 # max numbers of builds to keep
- artifactNumToKeepStr: 5 # number of builds that have their artifacts kept

Let's keep in sync.

File: ci/Jenkinsfiles/build.groovy [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+    stage('Build') {
+      steps {
+        container(DEFAULT_CONTAINER) {
+          setGitHubBuildStatus('maven/build', 'Build', 'PENDING')
+          sh "mvn ${MAVEN_ARGS} -V -T4C -DskipTests install"
+        }
+      }
+      post {
+        success {
+          archiveArtifacts allowEmptyArchive: true, artifacts: "${NUXEO_PACKAGE_PATH}"
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: Let's only archive the package if not a PR.

**Review by NourNuxeo — COMMENTED** [review]
👍  I'm more of a watcher on this PR. 
Don't hesitate to request me again as you go through the PR issues.

File: ci/Jenkinsfiles/build.groovy (hunk 8 above)
Comment (tfowley) [reply ci/Jenkinsfiles/build.groovy]: Did I add the block in the correct place here?

File: ci/Jenkinsfiles/build.groovy (hunk 10 above)
Comment (tfowley) [reply ci/Jenkinsfiles/build.groovy]: Just checking if I implemented the conditional correctly here

File: ci/Jenkinsfiles/build.groovy (hunk 8 above)
Comment (ataillefer) [reply ci/Jenkinsfiles/build.groovy]: Nope, we want it at the very beginning of the Groovy script, aka after the copyright.
basically, really copy:
```
/*
 * (C) Copyright 2019 Nuxeo (http://nuxeo.com/) and others.
 *
 * ...
 */
def abortRunningBuilds() {
  // see https://issues.jenkins.io/browse/JENKINS-43353
  def buildNumber = BUILD_NUMBER as int
  if (buildNumber > 1) {
    milestone(buildNumber - 1)
  }
  milestone(buildNumber)
}
abortRunningBuilds()
```

File: ci/Jenkinsfiles/build.groovy (hunk 10 above)
Comment (ataillefer) [reply ci/Jenkinsfiles/build.groovy]: Nope, `when` can only be used directly within `stage`, see https://www.jenkins.io/doc/book/pipeline/syntax/#when.
So, you should use script and if.

Comment (tfowley) [reply ci/Jenkinsfiles/build.groovy]: I had a suspicion but wanted to see if I could check it without script 😅

File: ci/Jenkinsfiles/build.groovy (hunk 8 above)
Comment (tfowley) [reply ci/Jenkinsfiles/build.groovy]: All good now?

Comment (ataillefer) [reply ci/Jenkinsfiles/build.groovy]: yep

File: ci/Jenkinsfiles/build.groovy [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+      steps {
+        container(DEFAULT_CONTAINER) {
+          setGitHubBuildStatus('maven/build', 'Build', 'PENDING')
+          sh "mvn ${MAVEN_ARGS} -V -T4C -DskipTests install"
+        }
+      }
+      post {
+        success {
+          script {
+            if(!isPullRequest){
```
Comment (ataillefer) [inline ci/Jenkinsfiles/build.groovy]: isPulllRequest() -> it's a function

**Review by ataillefer — COMMENTED** [review]
Any reason to separate the commits? I would say that removing the patch version (extra .0) from the POM files is part of the build process (related to the fixed version computation).

Also, for the commit message, I would keep it short: "NXP-30864: Implement CI build pipeline", we know its' for nuxeo-compound-documentq which is the current repository.

File: ci/Jenkinsfiles/build.groovy (hunk 11 above)
Comment (tfowley) [reply ci/Jenkinsfiles/build.groovy]: Whoops! 🤦‍♂️

**Comment by tfowley** [comment]
> Any reason to separate the commits? I would say that removing the patch version (extra .0) from the POM files is part of the build process (related to the fixed version computation).
> 
> Also, for the commit message, I would keep it short: "[NXP-30864](https://jira.nuxeo.com/browse/NXP-30864): Implement CI build pipeline", we know its' for nuxeo-compound-documentq which is the current repository.

Yes I wasn't 100% sure about the separate commit because it seemed potentially distinct enough to merit a "cleanup" type commit but having them all in the one Ci pipeline commit also makes sense to me. 
Will change the name and squash now 👍

**Review by ataillefer — APPROVED** [review]

**Review by troger — APPROVED** [review]

**Pull request merged**